examples/02_intermediate/wikimedia.ipynb

```diff
@@ -0,0 +1,513 @@
+   "source": [
+    "Next, we will need to write a new CSP [Adapter](https://github.com/Point72/csp/wiki/Adapters) to fetch the data and process it. In this example, we will write both a [Realtime Adapter](https://github.com/Point72/csp/wiki/Write-Realtime-Input-Adapters) and a [Historical Adapter](https://github.com/Point72/csp/wiki/Write-Historical-Input-Adapters)."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "8329b504-5699-48e4-9b14-fe647fccf48a",
+   "metadata": {},
+   "source": [
+    "## Writing adapters"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "5cfef5ff-9ddc-47cb-be18-ddb2d14b7446",
+   "metadata": {},
+   "source": [
+    "There are two main categories of writing input adapters, historical and realtime. When writing historical adapters you will need to implement a \"pull\" adapter, which pulls data from a historical data source in time order, one event at a time. When writing realtime adapters, you will need to implement a \"push\" adapter, which will get data from a separate thread that drives external events and \"pushes\" them into the engine as they occur.\n",
+    "\n",
+    "When writing input adapters it is also very important to denote the difference between \"graph building time\" and \"runtime\" versions of your adapter. Graph build time components solely describe the adapter. They are meant to do little else than keep track of the type of adapter and its parameters, which will then be used to construct the actual adapter implementation when the engine is constructed from the graph description. It is the runtime implementation that actual runs during the engine execution phase to process data."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "7922e29d-ebc5-443c-a0f3-7665a267f2c0",
+   "metadata": {},
+   "source": [
+    "## Using CSP to read realtime data"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "580a80bf-b012-4c5f-b5d9-e12a1a2d3909",
+   "metadata": {},
+   "source": [
+    "The data we are going to explore is the [MediaWiki Recent Changes feed](https://www.mediawiki.org/wiki/Manual:RCFeed) stream, which emits events related to recent changes across Wikimedia sites. The stream can be accessed through the [EventStreams web service](https://wikitech.wikimedia.org/wiki/Event_Platform/EventStreams_HTTP_Service)."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "5c4633f0-e2de-4bb0-9828-c6089b832e21",
+   "metadata": {},
+   "source": [
+    "Initially, we'll set the URL and define the data structure we expect to encounter. We will use a pandas DataFrame to manipulate the data."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "c4008a7b-36d6-4d9a-accc-379fca61236d",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import pandas as pd\n",
+    "from datetime import datetime\n",
+    "\n",
+    "URL = 'https://stream.wikimedia.org/v2/stream/recentchange'\n",
+    "\n",
+    "# Initial data\n",
+    "data = pd.DataFrame({\n",
```

Comment:
not following where this is used.  we should avoid using pandas here it can be more confusing that its worth

Reply:
I still think this pandas bit needs to be removed, shouldnt use pandas in this example anywhere

Reply:
The pandas use here is to prepare the Perspective widget. I will try to find a way to not use a DataFrame for this.